pkg/ds/product_left.go

```diff
@@ -0,0 +1,261 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package ds
+
+// ProductLeft is a cartesian product of two sets
+// The implementation represents the sets succinctly, merging keys with equivalent values to a single key,
+// so the mapping is injective (one-to-one).
+type ProductLeft[K Set[K], V Set[V]] struct {
+	m *HashMap[K, V]
+}
+
+// NewProductLeft with parameters [K, V] creates an empty Product[K, V] object,
+// implemented using K sets are keys and V sets as values.
+func NewProductLeft[K Set[K], V Set[V]]() *ProductLeft[K, V] {
+	return &ProductLeft[K, V]{m: NewHashMap[K, V]()}
+}
+
+// CartesianPairLeft returns a new Product object holding the cartesian product of the input sets k x v.
+// If either k or v is empty, the result is an empty Product object.
+func CartesianPairLeft[K Set[K], V Set[V]](k K, v V) *ProductLeft[K, V] {
+	m := NewProductLeft[K, V]()
+	if !k.IsEmpty() && !v.IsEmpty() {
+		m.m.Insert(k, v)
+	}
+	return m
+}
+
+// Left returns the projection Product[K, V] on the set K.
+func (m *ProductLeft[K, V]) Left(empty K) K {
+	res := empty.Copy()
+	for _, p := range m.Partitions() {
+		res = res.Union(p.Left)
+	}
+	return res
+}
+
+// Right returns the projection Product[K, V] on the set V.
+func (m *ProductLeft[K, V]) Right(empty V) V {
+	res := empty.Copy()
+	for _, p := range m.Partitions() {
+		res = res.Union(p.Right)
+	}
+	return res
+}
+
+func asLeftProduct[K Set[K], V Set[V]](m Product[K, V]) *ProductLeft[K, V] {
+	p, ok := m.(*ProductLeft[K, V])
+	if ok {
+		return p
+	}
+	res := NewProductLeft[K, V]()
+	for _, pair := range m.Partitions() {
+		res.m.Insert(pair.Left, pair.Right)
+	}
+	res.canonicalize()
+	return res
+}
+
+// Equal returns true if this and other are equivalent Product object.
+func (m *ProductLeft[K, V]) Equal(other Product[K, V]) bool {
+	return m.m.Equal(asLeftProduct(other).m)
+}
+
+// Copy returns a deep copy of the Product object.
+func (m *ProductLeft[K, V]) Copy() Product[K, V] {
+	return &ProductLeft[K, V]{m: m.m.Copy()}
+}
+
+// Hash returns the hash value of the Product object
+func (m *ProductLeft[K, V]) Hash() int {
+	const rrr = 5
+	res := rrr
+	for _, p := range m.Partitions() {
+		res ^= (p.Left.Hash() << 1) ^ p.Right.Hash()
+	}
+	return res
+}
+
+// IsEmpty returns true if the Product object is empty.
+func (m *ProductLeft[K, V]) IsEmpty() bool {
+	return m.m.IsEmpty()
+}
+
+// Size returns the number of unique pairs in the Product object
```

Comment:
the Size documentation is incorrect? it does not return the number of pairs..

Reply:
I'm not sure I understand. Maybe the issue here is the notion of "pair" - here it means the mathematical pair of underlying elements, [x, y], not a pair of sets [s1, s2] (usually implicitly encoding s1xs2, as is the case in the return value of `Partitions()`.